async/Mailbox.go

```diff
@@ -0,0 +1,119 @@
+// A value durably.  A value is considered durably stored if it successfully
+// writes to two of three replicas. We want to write to all replicas in parallel
+// and return as soon as two writes succeed. We return an error if < 2 writes succeed
+//
+//  func storeValue(num int) error {
+//    successfulWrites := 0
+//    returnedWrites := 0
+//    mailbox := NewAsyncMailbox()
+//
+//    writeCallback := func (err error) {
+//      if err != nil {
+//        successfulWrites++
+//      }
+//      returnedWrites++
+//    }
+//
+//    // Send to Replica One
+//    go func(rsp *AsyncError){
+//      rsp.SetValue(write(num, "replicaOne"))
+//    }(mailbox.NewAsyncError(writeCallback))
+//
+//    // Send to Replica Two
+//    go func(rsp *AsyncError){
+//      rsp.SetValue(write(num, "replicaTwo"))
+//    }(mailbox.NewAsyncError(writeCallback))
+//
+//    // Send to Replica Three
+//    go func(rsp *AsyncError){
+//      rsp.SetValue(write(num, "replicaThree"))
+//    }(mailbox.NewAsyncError(writeCallback))
+//
+//    // Value is Considered Durably Stored if at least two write calls succeeded
+//    for sucessfullWrites < 2 && returnedWrites < 3 {
+//       mailbox.ProcessMessages()
+//    }
+//
+//    if successfulWrites >= 2 {
+//      return nil
+//    } else {
+//      return errors.New("Could Not Durably Store Value")
+//    }
+//
+//  // a function which makes a call to a durable register
+//  // which is accessed via the network
+//  func write (num int, address string) error { ... }
+//
+// A Mailbox is not a concurrent structure and should only
+// ever be accessed from a single go routine.  This ensures that the callbacks
+// are always executed within the same context and only one at at ime.
+// A Mailbox for keeping track of in progress AsyncMessages.
+// This structure is not thread-safe.
+type Mailbox struct {
+	msgs []message
+}
+
+// The function type of the callback invoked when an AsyncError is Completed
+type AsyncErrorResponseHandler func(error)
+
+// async message is a struct composed of an AsyncError
+// and its associated callback
+type message struct {
+	Err      *AsyncError
+	callback AsyncErrorResponseHandler
+}
+
+func newMessage(cb AsyncErrorResponseHandler) message {
+	return message{
+		Err:      newAsyncError(),
+		callback: cb,
+	}
+}
+
+func NewMailbox() *Mailbox {
+	return &Mailbox{
+		msgs: make([]message, 0),
+	}
+}
+
+// Creates a NewAsyncError and assocaites the supplied callback with it.
+// Once the AsyncError has been completed, SetValue called, the callback
+// will be invoked on the next execution of ProcessMessages
+func (bx *Mailbox) NewAsyncError(cb AsyncErrorResponseHandler) *AsyncError {
+	msg := newMessage(cb)
+	bx.msgs = append(bx.msgs, msg)
+	return msg.Err
+}
+
+// Processes the mailbox.  For all messages with completed AsyncErrors
+// the callback function and removes the message from the mailbox
+func (bx *Mailbox) ProcessMessages() {
```

Comment:
It's still unclear to me if scoot will need multiple mailboxes or if not, how sibling components can interact with a single mailbox. It seems like we'd need locking for ProcessMessages() and NewAsyncError() so we can pass mailbox around to different funcs and goroutines.

I'm still not a big fan of AsyncError even though I completely see why it makes sense now. Maybe ErrorPromise since that emphasizes that we need to fulfill that promise and await on it, whereas Async relates more to the concept of concurrency and goroutines.